lib/Sema/CSApply.cpp

```diff
@@ -4241,6 +4234,20 @@ namespace {
           component = KeyPathExpr::Component::forProperty(ref,
                                                        resolvedTy,
                                                        origComponent.getLoc());
+
+          baseTy = component.getComponentType();
+          resolvedComponents.push_back(component);
+
+          if (shouldForceUnwrapResult(property, locator)) {
+            assert(!baseTy->hasTypeVariable());
+            auto objTy = baseTy->getRValueType()->getAnyOptionalObjectType();
```

Comment:
Can you factor this out into a new function or method, since it appears below at line 4322 also?

Reply:
I have another branch where I am experimenting with LValues and Optionals and adding a helper to make this code and other code in the solver simpler. If we had a helper that produced an LValue(Object) from LValue(Optional(Object)) I don't think it would make much sense to have a separate function for this.

Reply:
I ended up refactoring the lvalue/optional part of this into a lambda and use that in all the places this is done (not just the couple I added) throughout this function. I'll push that version up once I think it's going to pass at least as many tests. Still trying to figure out what's going on on Linux.